hal/inc/ble_hal.h

```diff
@@ -15,190 +15,747 @@
+    uint16_t char_handle;
+    uint8_t* descriptor;
+    uint16_t len;
+    hal_ble_uuid_t uuid;
+} hal_ble_desc_init_t;
+
+/* BLE events structure */
+typedef struct {
+    uint8_t version;
+    void* reserved;
+} hal_ble_gap_on_adv_stopped_evt_t;
+
+typedef struct {
+    uint8_t version;
+    int8_t rssi;
+    struct {
+        uint16_t connectable   : 1;     /**< Connectable advertising event type. */
+        uint16_t scannable     : 1;     /**< Scannable advertising event type. */
+        uint16_t directed      : 1;     /**< Directed advertising event type. */
+        uint16_t extended_pdu  : 1;     /**< Received an extended advertising set. */
+    } type;
+    uint16_t adv_data_len;
+    uint8_t* adv_data;
+    uint16_t sr_data_len;
+    uint8_t* sr_data;
+    hal_ble_addr_t peer_addr;
+} hal_ble_gap_on_scan_result_evt_t;
+
+typedef struct {
+    uint8_t version;
+    uint8_t role;
+    uint16_t conn_handle;
+    uint16_t conn_interval;             /**< Connection Interval in 1.25 ms units.*/
+    uint16_t slave_latency;             /**< Slave Latency in number of connection events.*/
+    uint16_t conn_sup_timeout;          /**< Connection Supervision Timeout in 10 ms units.*/
+    hal_ble_addr_t peer_addr;
+} hal_ble_gap_on_connected_evt_t;
+
+typedef struct {
+    uint8_t version;
+    uint8_t reason;
+    uint16_t conn_handle;
+} hal_ble_gap_on_disconnected_evt_t;
 
-// API version
-#define BLE_API_VERSION 1
+typedef struct {
+    uint8_t version;
+    uint16_t conn_handle;
+    uint16_t conn_interval;             /**< Connection Interval in 1.25 ms units.*/
+    uint16_t slave_latency;             /**< Slave Latency in number of connection events.*/
+    uint16_t conn_sup_timeout;          /**< Connection Supervision Timeout in 10 ms units.*/
+} hal_ble_gap_on_conn_params_evt_t;
 
-// Default maximum size of an ATT packet in bytes (ATT_MTU)
-#define BLE_MIN_ATT_MTU_SIZE 23
+typedef struct {
+    uint8_t version;
+    uint16_t conn_handle;
+    uint16_t att_mtu_size;
+} hal_ble_gatt_on_params_updated_evt_t;
 
-// Size of the ATT opcode field in bytes
-#define BLE_ATT_OPCODE_SIZE 1
+typedef struct {
+    uint8_t version;
+    uint16_t conn_handle;
+    uint8_t  count;
+    hal_ble_svc_t* services;
+} hal_ble_gattc_on_svc_disc_evt_t;
 
-// Size of the ATT handle field in bytes
-#define BLE_ATT_HANDLE_SIZE 2
+typedef struct {
+    uint8_t version;
+    uint16_t conn_handle;
+    uint8_t count;
+    hal_ble_char_t* characteristics;
+} hal_ble_gattc_on_char_disc_evt_t;
 
-// Minimum and maximum number of bytes that can be sent in a single write command, read response,
-// notification or indication packet
-#define BLE_MIN_ATTR_VALUE_PACKET_SIZE (BLE_MIN_ATT_MTU_SIZE - BLE_ATT_OPCODE_SIZE - BLE_ATT_HANDLE_SIZE)
-#define BLE_MAX_ATTR_VALUE_PACKET_SIZE (BLE_MAX_ATT_MTU_SIZE - BLE_ATT_OPCODE_SIZE - BLE_ATT_HANDLE_SIZE)
+typedef struct {
+    uint8_t version;
+    uint16_t conn_handle;
+    uint16_t attr_handle;
+    uint16_t offset;
```

Comment:
Consider using larger data type (perhaps `size_t`) here as ATT MTU is not limited by 65535 by the standard.